CRM/Mailing/BAO/MailingJob.php

```diff
@@ -831,11 +831,54 @@ public static function cancel($mailingId) {
         2 => array(date('YmdHis'), 'Timestamp'),
       );
       CRM_Core_DAO::executeQuery($sql, $params);
-
-      CRM_Core_Session::setStatus(ts('The mailing has been canceled.'), ts('Canceled'), 'success');
     }
   }
 
+  /**
+   * Pause a mailing
+   *
+   * @param int $mailingID
+   *   The id of the mailing to be paused.
+   */
+  public static function pause($mailingID) {
+    $sql = "
+      UPDATE civicrm_mailing_job
+      SET status = 'Paused'
+      WHERE mailing_id = %1
+      AND is_test = 0
+      AND status IN ('Scheduled', 'Running')
```

Comment:
For `Scheduled` jobs, switching to `Paused` seems pretty straight-forward.

For `Running` jobs, I think it's less obvious. It's also pretty hard to test due to the intrinsic raciness. First... what do we expect to happen?

* A typical user probably expects an  *instantaneous pause*: you hit the `Pause` on a media player (movie/music), and there is a a total and immediate cessation of active work.
* As someone who's done development with CiviMail, I expect a *gradual pause*: you hit the breaks on a car/bus/bike, and inertia carries you a bit further. If there's an active-delivery of a batch of messages, that continues. But new batches won't be started. *However*, this depends a lot on the local setup. If you have several concurrent workers, or if batching is effectively disabled (high or non-existent limits), then it may not feel like `Pause` accomplishes anything.

IMHO, the *gradual pause* is fine. It's a lot better than what we had before. And the main audience which needs `Pause` would be folks who send large blasts... those are the same folks who need batching.

But there are two concerns coming out of this:

 * There should be documentation or communication somewhere which clarifies that `Pause` is gradual, not instantaneous -- and that it's dependent on your batch settings. For example:
    * In the `setStatus()` communication, maybe say, "The mailing has been paused. Active message deliveries may continue for a few minutes, but CiviMail will not begin delivery of any more batches."
    * Or find somewhere else to put documentation...
 * There are several different batch configurations (e.g. with subjobs; e.g. with per-cron-run limits; e.g. with 1 or multiple workers) and a few moments at which one might stop delivery (e.g. during the first batch; during a middle batch; during the last/only batch; e.g. repeatedly pausing/unpausing in rapid succession), and we need some empirical basis for saying that it works in various permutations. It doesn't have to be a formal unit-test for every scenario, and it doesn't have to be an independent reviewer testing each of these. @jitendrapurohit if you run through a few scenarios (or already have run), just report back on which scenarios were tested.

Reply:
This patch is applied on one of our live site from around a week and we haven't seen any issues related to this. Agree with the documentation and happy to update the sucess message as above. For the scenarios tested by Fuzion -

We have **only** tested this for the simplest case with one thread/worker.

We have paused and restarted in between batches with success. Not paused whilst the batch was actually running (ie. a long running batch).